Automattic/wp-calypso
# Free Trials: Move free trial checkout to a new controller

Move the free trial checkout to its own controller. This has the advantage of lightening the checkout component complexity and fixing a few minor bugs which made the free trial upgrade conflict with other cart items.

#### Context
See this comment https://github.com/Automattic/wp-calypso/issues/286#issuecomment-171041340

Should fix #1538 and #2207 and #2274 as well.

### Testing instructions
`git checkout update/free-trial-checkout-page`

#### Test 1
- Navigate to http://calypso.localhost:3000/plans/:site for a site eligible to free trial
- Click on `Start Free Trials` and check that you are redirected to the Free Trial confirmation page.
- Confirm subscription and check that you are redirected to the Plans page and that the thank you notice is displayed.
- Close the notice

#### Test 2 (needs a fix in wpcom)
- Add a product (such as a premium theme) to your cart.
- Don't complete checkout yet but go to the plans page and start a free trial.
- Complete checkout of your product.

#### Test 3
- Navigate to `/start/free-trial`, choose a theme and a name for your new site.
- On the Plans page click on one of the `Start Free Trials` button.
- Ensure you are redirected to `/plans/:site/thank-you` and that the thank you notice is displaying.

#### Test 4 (logged out)
- Navigate to `/start/free-trial`
- Enter `localStorage.setItem( 'ABTests', '{"freeTrials_20160112":"offered"}' );` in the console to add yourself to the test
- Choose a theme and a name for your new site.
- On the Plans page click on one of the `Start Free Trials` button.
- Choose a name and password for your new account.
- Confirm your free trial subscription
- Ensure that you are redirected to `/plans/:site/thank-you` and that the thank you notice is displaying (along with the email confirmation notice).

#### Test 5
- Cancel a free trial from the admin tools (Only available for automatticians)
- Go to the Plans page and check that the other plan is still available for a free trial.
- Click on `Start Free Trials`
- Change the url to the other plan and reload the page (for instance `/start-trial/premium/:site` becomes `/start-trial/business/:site`
- Assert that you have a message telling you you cannot purchase and inviting you to go back to the Plans page

#### Reviews
